### DISCUSSION TITLE
Adding userId / author to score (custom metadata)

### DISCUSSION THREAD

**Discussion by ivanviragine** [langfuse/langfuse#2469] [Ideas]
### Describe the feature or potential improvement

It would be nice to have the userId or other identifier on scores created via API. Now it accepts only:

```
{
  "id": "string",
  "traceId": "string",
  "name": "string",
  "value": 0,
  "observationId": "string",
  "comment": "string"
}
```

But on my app several users can actually evaluate a single generation (it's not like a private chat).

**Comment by marcklingen** [comment]
Is this the userId as in trace.userId?

**Reply by ivanviragine** [reply]
It's a way to tag the id who made the evaluation. Let's say that I have an application that generates answers for questions once they are asked. Let's say user A asked question X and it generates a trace. If another user asks the same question, the user B will see the same answer, but, since it's the same answer it'll not necessarily be a new trace or LLM call. Both user A and B (and any other) can evaluate the answer (that's on the trace).
Is it clearer now?

**Reply by marcklingen** [reply]
makes sense, currently there is an authorId on scores which relates scores to internal users (who have access to the langfuse instance). When filtering by userId (external user) scores get attributed to the userId on the trace to which the score is associated currently. This seems to be a sensible default for most applications.

as a workaround you could add the userId to the comment field of the score (you can filter by it) but this is not ideal

**Reply by ivanviragine** [reply]
The way it is now it makes sense in a chat or other applications which the traces are kinda related only to the user in a one-one relationship. I agree that's the majority of cases, but it would be nice to have that option too!

Thanks a lot.

**Reply by marcklingen** [reply]
we could also add an optional userId field which you could then override if it deviates from the trace.userId. This just makes ingesting scores into langfuse more expensive but could be a good tradeoff

**Reply by ivanviragine** [reply]
Of course since it's my main usage I would vote for that. But, I imagine that there are way more traces than evaluation, so a little overhead on it wouldn't hurt. 
But that's your call. 
Thank you very much for the attention!

**Reply by dsvolk** [reply]
I'd also like to have it.

In my flow, the evaluators for my LLM app answers are not necessary the same people who actually run the LLM app. Moreover, I also do batch runs across multiple inputs and then ask a few evaluators to score them, so I'd like to track who gave which score.

**Reply by marcklingen** [reply]
thanks for sharing @dsvolk! Does logging this information as a score.comment work for you as well as an intermediary solution?

**Reply by dsvolk** [reply]
Well a rather big problem with this solution is that I'd need to pull **all** my scores and parse them every time I need to query them in some way 🤔

**Comment by ivanviragine** [comment]
Any news on this? Thanks!

**Reply by marcklingen** [reply]
No changes on this yet, open for contributions or ideas

**Reply by ivanviragine** [reply]
Maybe the score config could provide a way to add custom fields, since it's a "schema-like" idea. that way it would be more flexible and won't need too much efforts on the implementation. What do you think?

**Reply by marcklingen** [reply]
That's a good one

alternatively, an open-ended metadata field (json) could solve for many use cases that are non-standard for most teams

**Reply by ivanviragine** [reply]
Either one of the two works for me!

**Comment by jpmin7** [comment]
I would also prefer the open-ended metadata field (json) since there is more metadata to add from the production context (specific to our use case) such as userId, projectId, sessionId. 

Also, then filtering this field in the UI would help a lot to understand user feedback better.

**Reply by marcklingen** [reply]
thanks for sharing, makes sense!
